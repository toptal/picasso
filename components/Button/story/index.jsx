import PicassoBook from '../../../.storybook/components/PicassoBook'

const page = PicassoBook.createPage('Button')

page
  .addExample('Button/story/Basic-example.jsx', 'Basic')
  .addExample('Button/story/Variants-example.jsx', 'Variants')
  .addExample('Button/story/States-example.jsx', 'States')
  .addExample('Button/story/Disabled-example.jsx', {
    title: 'Disabled',
    description: 'The button shows that currently unable to be interacted with'
  })
  .addExample('Button/story/Sizes-example.jsx', 'Sizes')
  .addExample('Button/story/FullWidth-example.jsx', 'Full width')
  .addExample('Button/story/IconButtons-example.jsx', 'Button with Icon')
  .addExample(
    'Button/story/IconButtonsWithText-example.jsx',
    'Button with text and Icon'
  )
  .addExample('Button/story/Loading-example.jsx', 'Button with loading state')
