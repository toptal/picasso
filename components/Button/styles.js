import { darken } from '../styles'

const ICON_SPACING = '0.4em'
const getFilledButton = (backgroundColor, borderColor, color) => ({
  backgroundColor,
  borderColor,
  color,

  '&:hover, &$hovered': {
    backgroundColor: darken(backgroundColor, 0.05)
  },

  '&:active, &$active': {
    backgroundColor: darken(backgroundColor, 0.25),
    borderColor: darken(backgroundColor, 0.25)
  }
})

export default ({ palette, typography, transitions }) => ({
  root: {
    position: 'relative',
    textTransform: 'none',
    fontSize: typography.button.fontSize,
    padding: '.4em 1em',
    borderRadius: '.25rem',
    lineHeight: '1.5em',
    transition: `all ${transitions.duration.short}ms ${
      transitions.easing.easeOut
    }`,
    transitionProperty: 'border, color, background',
    border: `solid 1px ${palette.grey[50]}`,
    backgroundColor: palette.grey[50],

    '&:hover, &$hovered': {
      backgroundColor: darken(palette.grey[50], 0.05)
    },

    '&[disabled]': {
      opacity: 0.45
    },

    '&:focus, &$focused': {
      textDecoration: 'underline',

      '&:active, &$active, &:hover': {
        textDecoration: 'none'
      }
    },

    '&:active, &$active': {
      backgroundColor: darken(palette.grey[50], 0.15)
    },

    '&+&': {
      marginLeft: '0.5em'
    }
  },
  loader: {
    position: 'absolute',
    left: '50%',
    top: '50%',
    transform: 'translate(-50%, -50%)'
  },

  // sizes
  small: {
    fontSize: '0.8rem'
  },
  large: {
    fontSize: '1.2rem'
  },

  // Variants
  primary: getFilledButton(
    palette.primary.main,
    palette.primary.main,
    palette.common.white
  ),
  secondary: {
    ...getFilledButton(
      'transparent',
      palette.primary.main,
      palette.primary.main
    ),

    '&:hover, &$hovered': {
      backgroundColor: palette.primary.light,
      borderColor: palette.primary.main
    },

    '&:active, &$active': {
      backgroundColor: darken(palette.primary.light, 0.25),
      borderColor: darken(palette.primary.light, 0.25)
    }
  },
  success: getFilledButton(
    palette.success.main,
    palette.success.main,
    palette.common.white
  ),
  error: getFilledButton(
    palette.error.main,
    palette.error.main,
    palette.common.white
  ),
  flat: {
    ...getFilledButton(
      palette.common.white,
      palette.grey[50],
      palette.text.primary
    ),
    border: 'none'
  },
  basic: getFilledButton(
    palette.common.white,
    palette.grey[50],
    palette.text.primary
  ),

  // Other props
  fullWidth: {
    width: '100%'
  },
  hovered: {},
  focused: {},
  active: {},

  // Child elements
  icon: {
    fontSize: '1.2em !important',
    marginTop: '0.1em',
    verticalAlign: 'top'
  },
  iconLeft: {
    marginRight: ICON_SPACING
  },
  iconRight: {
    marginLeft: ICON_SPACING
  },
  children: {
    display: 'inline-block'
  },
  hidden: {
    opacity: 0
  }
})
