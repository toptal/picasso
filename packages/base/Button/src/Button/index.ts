import type { Props } from './Button'
export { default as Button } from './Button'
export type ButtonProps = Props
export * from './Button'
export * from './styles'
