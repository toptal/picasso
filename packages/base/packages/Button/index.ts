/* eslint-disable import/no-extraneous-dependencies */
import type { Props } from './Button'
export { default } from './Button'
export type ButtonProps = Props
export * from './Button'
