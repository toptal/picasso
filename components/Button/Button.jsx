import Button from '@material-ui/core/Button'
import { withStyles } from '@material-ui/core/styles'

export default withStyles({
  root: {
    textTransform: 'none',
    padding: '6px 16px',
    fontSize: '16px'
  },
  containedPrimary: {
    color: 'white',
    backgroundColor: '#204ecf',

    '&:hover': {
      backgroundColor: '#1542c1'
    }
  },
  outlinedSecondary: {
    color: '#204ecf',
    borderColor: '#204ecf',

    '&:hover': {
      backgroundColor: '#cad5f4',
      borderColor: '#204ecf'
    }
  }
})(Button)
