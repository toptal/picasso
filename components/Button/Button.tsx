import React, {
  FunctionComponent,
  ReactNode,
  ReactElement,
  MouseEvent
} from 'react'
import cx from 'classnames'
import { withStyles } from '@material-ui/core/styles'
import ButtonBase from '@material-ui/core/ButtonBase'

import { Classes } from '../styles/types'
import Loader from '../Loader'
import Group from '../ButtonGroup'
import styles from './styles'

type VariantType =
  | 'primary'
  | 'secondary'
  | 'flat'
  | 'basic'
  | 'success'
  | 'error'
  | 'default'

type SizeType = 'small' | 'medium' | 'large'

type IconPositionType = 'left' | 'right'

interface StylesProps {
  classes: Classes
}

interface Props {
  /** Show button in the active state (left mouse button down) */
  active?: boolean
  /** Content of Button component */
  children?: ReactNode
  className?: string
  focused?: boolean
  /** Take the full width of a container */
  fullWidth?: boolean
  hovered?: boolean
  /** Add an `<Icon />` along Button's children */
  icon?: ReactElement
  /** Icon can be positioned on the left or right */
  iconPosition?: IconPositionType
  /** A button can show a loading indicator */
  loading?: boolean
  /** Callback invoked when component is clicked */
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void
  /** A button can have different sizes */
  size?: SizeType
  /** The variant to use */
  variant?: VariantType
}

interface StaticProps {
  Group: typeof Group
}

export const Button: FunctionComponent<Props & StylesProps> &
StaticProps = props => {
  const {
    icon,
    iconPosition,
    loading,
    children,
    classes,
    className,
    fullWidth,
    variant,
    size,
    focused,
    hovered,
    active,
    onClick,
    ...rest
  } = props
  const {
    icon: iconClass,
    iconLeft: iconLeftClass,
    iconRight: iconRightClass,
    root: rootClass,
    children: childrenClass,
    hidden: hiddenClass,
    loader: loaderClass
  } = classes

  let finalChildren = [children]

  if (icon) {
    const iconComponent = React.cloneElement(icon, {
      className: cx(iconClass, {
        [iconLeftClass]: children && iconPosition === 'left',
        [iconRightClass]: children && iconPosition === 'right'
      }),
      key: 'button-icon'
    })

    if (iconPosition === 'left') {
      finalChildren.unshift(iconComponent)
    } else {
      finalChildren.push(iconComponent)
    }
  }

  const variantClassName = classes[variant!]
  const sizeClassName = classes[size!]

  const rootClassName = cx(
    {
      [classes.fullWidth]: fullWidth,
      [classes.active]: active,
      [classes.focused]: focused,
      [classes.hovered]: hovered
    },
    variantClassName,
    sizeClassName,
    rootClass
  )

  return (
    <ButtonBase
      classes={{
        root: rootClassName
      }}
      onClick={onClick}
      className={className}
      {...rest}
    >
      <div className={cx(childrenClass, className, { [hiddenClass]: loading })}>
        {finalChildren}
      </div>

      {loading && <Loader className={loaderClass} inline size='small' />}
    </ButtonBase>
  )
}

Button.defaultProps = {
  active: false,
  children: null,
  focused: false,
  fullWidth: false,
  hovered: false,
  iconPosition: 'left',
  loading: false,
  onClick: () => {},
  size: 'medium',
  variant: 'default'
}

Button.displayName = 'Button'

Button.Group = Group

export default withStyles(styles)(Button) as FunctionComponent<Props> &
  StaticProps
