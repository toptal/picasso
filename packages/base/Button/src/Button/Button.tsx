import type { ReactNode, ReactElement, MouseEvent, ElementType } from 'react'
import React, { forwardRef } from 'react'
import cx from 'classnames'
import type {
  StandardProps,
  SizeType,
  ButtonOrAnchorProps,
  OverridableComponent,
  TextLabelProps,
  Classes,
} from '@toptal/picasso-shared'
import { useTitleCase } from '@toptal/picasso-shared'
import type { Theme } from '@material-ui/core'
import { makeStyles, ButtonBase } from '@material-ui/core'
import { Loader } from '@toptal/picasso-loader'
import { Container } from '@toptal/picasso-container'
import { noop, toTitleCase } from '@toptal/picasso-utils'
// we need to ensure the correct order of styles import
// TODO: [FX-4614] To be removed when Link component is migrated to tailwind
import { Link } from '@toptal/picasso-link'

import styles from './styles'

// HACK: This statement is only used to prevent webpack from tree shaking the import
void Link

const useStyles = makeStyles<Theme, Props>(styles, {
  name: 'PicassoButton',
})

export type VariantType =
  | 'primary'
  | 'negative'
  | 'positive'
  | 'secondary'
  | 'transparent'

export type IconPositionType = 'left' | 'right'

export interface Props
  extends StandardProps,
    TextLabelProps,
    ButtonOrAnchorProps {
  /** Show button in the active state (left mouse button down) */
  active?: boolean
  as?: ElementType
  /** Disables button */
  disabled?: boolean
  /** Content of Button component */
  children?: ReactNode
  focused?: boolean
  /** Take the full width of a container */
  fullWidth?: boolean
  /** Set hovered style for the button */
  hovered?: boolean
  /** Add an `<Icon />` along Button's children */
  icon?: ReactElement
  /** Icon can be positioned on the left or right */
  iconPosition?: IconPositionType
  /** Shows a loading indicator and disables click events */
  loading?: boolean
  /** Callback invoked when component is clicked */
  onClick?: (event: MouseEvent<HTMLButtonElement & HTMLAnchorElement>) => void
  /** A button can have different sizes */
  size?: SizeType<'small' | 'medium' | 'large'>
  /** The variant to use */
  variant?: VariantType
  /** HTML Value of Button component */
  value?: string | number
  /** HTML title of Button component */
  title?: string
  /** HTML type of Button component */
  type?: 'button' | 'reset' | 'submit'
}

const getClickHandler = (loading?: boolean, handler?: Props['onClick']) =>
  loading ? noop : handler

const getIcon = (
  classes: Classes,
  children: ReactNode,
  icon?: ReactElement,
  iconPosition?: IconPositionType
) => {
  if (!icon) {
    return null
  }

  const {
    icon: iconClass,
    iconLeft: iconLeftClass,
    iconRight: iconRightClass,
  } = classes

  return React.cloneElement(icon, {
    className: cx(iconClass, icon.props.className, {
      [iconLeftClass]: children && iconPosition === 'left',
      [iconRightClass]: children && iconPosition === 'right',
    }),
    key: 'button-icon',
  })
}

export const Button: OverridableComponent<Props> = forwardRef<
  HTMLButtonElement,
  Props
>(function Button(props, ref) {
  const {
    icon,
    iconPosition,
    loading,
    children,
    className,
    style,
    fullWidth,
    variant = 'primary',
    size = 'medium',
    focused,
    hovered,
    disabled,
    active,
    onClick,
    title,
    value,
    type,
    as = 'button',
    titleCase: propsTitleCase,
    ...rest
  } = props
  const classes = useStyles(props)

  const {
    root: rootClass,
    hidden: hiddenClass,
    loader: loaderClass,
    content: contentClass,
  } = classes

  const titleCase = useTitleCase(propsTitleCase)

  const finalChildren = [titleCase ? toTitleCase(children) : children]

  if (icon) {
    const iconComponent = getIcon(classes, children, icon, iconPosition)

    if (iconPosition === 'left') {
      finalChildren.unshift(iconComponent)
    } else {
      finalChildren.push(iconComponent)
    }
  }

  const variantClassName = classes[variant]
  const sizeClassName = classes[size]

  const rootClassName = cx(
    {
      [classes.fullWidth]: fullWidth,
      [classes.active]: active,
      [classes.focused]: focused,
      [classes.hovered]: hovered,
      [classes.disabled]: disabled,
    },
    sizeClassName,
    variantClassName,
    rootClass
  )

  return (
    <ButtonBase
      {...rest}
      ref={ref}
      classes={{
        root: rootClassName,
        focusVisible: cx(classes.focusVisible),
      }}
      onClick={getClickHandler(loading, onClick)}
      className={className}
      style={style}
      disabled={disabled}
      title={title}
      value={value}
      type={type}
      component={as}
      data-component-type='button'
    >
      <Container
        as='span'
        inline
        flex
        direction='row'
        alignItems='center'
        className={cx({ [hiddenClass]: loading }, contentClass)}
      >
        {finalChildren}
      </Container>

      {loading && (
        <Loader variant='inherit' className={loaderClass} inline size='small' />
      )}
    </ButtonBase>
  )
})

Button.defaultProps = {
  active: false,
  as: 'button',
  children: null,
  disabled: false,
  focused: false,
  fullWidth: false,
  hovered: false,
  iconPosition: 'left',
  loading: false,
  onClick: noop,
  size: 'medium',
  type: 'button',
  variant: 'primary',
}

Button.displayName = 'Button'

export default Button
