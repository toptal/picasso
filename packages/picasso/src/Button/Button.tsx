import React, {
  ReactNode,
  ReactElement,
  MouseEvent,
  forwardRef,
  ElementType
} from 'react'
import cx from 'classnames'
import { makeStyles, Theme } from '@material-ui/core/styles'
import ButtonBase from '@material-ui/core/ButtonBase'
import {
  BaseProps,
  SizeType,
  ButtonOrAnchorProps,
  CompoundedComponentWithRef,
  OverridableComponent,
  useAppConfig
} from '@toptal/picasso-shared'

import Loader from '../Loader'
import Container from '../Container'
import Group from '../ButtonGroup'
import kebabToCamelCase from '../utils/kebab-to-camel-case'
import toTitleCase from '../utils/to-title-case'
import styles from './styles'

export type VariantType =
  | 'primary-blue'
  | 'secondary-blue'
  | 'primary-red'
  | 'secondary-red'
  | 'primary-green'
  | 'secondary-green'
  | 'flat'
  | 'flat-white'
  | 'secondary-white'
  | 'transparent'
  | 'transparent-white'
  | 'transparent-blue'

export type IconPositionType = 'left' | 'right'

export interface Props extends BaseProps, ButtonOrAnchorProps {
  /** Show button in the active state (left mouse button down) */
  active?: boolean
  /** The component used for the root node. Either a string to use a DOM element or a component. */
  as?: ElementType
  /** Disables button */
  disabled?: boolean
  /** Content of Button component */
  children?: ReactNode
  // TODO: should it be exposed?
  focused?: boolean
  /** Take the full width of a container */
  fullWidth?: boolean
  /** Set hovered style for the button */
  hovered?: boolean
  /** Add an `<Icon />` along Button's children */
  icon?: ReactElement
  /** Icon can be positioned on the left or right */
  iconPosition?: IconPositionType
  /** Shows a loading indicator and disables click events */
  loading?: boolean
  /** Callback invoked when component is clicked */
  onClick?: (event: MouseEvent<HTMLButtonElement & HTMLAnchorElement>) => void
  /** A button can have different sizes */
  size?: SizeType<'small' | 'medium' | 'large'>
  /** The variant to use */
  variant?: VariantType
  /** HTML Value of Button component */
  value?: string | number
  /** Circular style of Button component */
  circular?: boolean
  /** HTML title of Button component */
  title?: string
  /** HTML type of Button component */
  type?: 'button' | 'reset' | 'submit'
}

interface StaticProps {
  Group: typeof Group
}

const getVariantType = (variant: VariantType) => {
  const [type] = variant!.split('-')

  return type
}

const useStyles = makeStyles<Theme, Props>(styles, { name: 'PicassoButton' })

const defaultOnClick = () => {}

export const Button = forwardRef<HTMLButtonElement, Props>(function Button(
  props,
  ref
) {
  const {
    icon,
    iconPosition,
    loading,
    children,
    className,
    style,
    fullWidth,
    variant,
    size,
    focused,
    hovered,
    disabled,
    active,
    onClick,
    circular,
    title,
    value,
    type,
    as,
    ...rest
  } = props
  const classes = useStyles(props)

  const { titleCase } = useAppConfig()

  const {
    icon: iconClass,
    iconLeft: iconLeftClass,
    iconRight: iconRightClass,
    root: rootClass,
    hidden: hiddenClass,
    loader: loaderClass,
    content: contentClass
  } = classes

  const finalChildren = [titleCase ? toTitleCase(children) : children]

  if (icon) {
    const iconComponent = React.cloneElement(icon, {
      className: cx(iconClass, icon.props.className, {
        [iconLeftClass]: children && iconPosition === 'left',
        [iconRightClass]: children && iconPosition === 'right'
      }),
      key: 'button-icon'
    })

    if (iconPosition === 'left') {
      finalChildren.unshift(iconComponent)
    } else {
      finalChildren.push(iconComponent)
    }
  }

  const variantType = getVariantType(variant!)
  const variantClassName = disabled
    ? classes[`${variantType}Disabled`]
    : classes[kebabToCamelCase(variant!)]
  const sizeClassName = classes[size!]

  const rootClassName = cx(
    {
      [classes.fullWidth]: fullWidth,
      [classes.active]: active,
      [classes.focused]: focused,
      [classes.hovered]: hovered,
      [classes.circular]: circular
    },
    sizeClassName,
    variantClassName,
    rootClass
  )

  return (
    <ButtonBase
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
      ref={ref}
      classes={{
        root: rootClassName
      }}
      onClick={loading ? defaultOnClick : onClick}
      className={className}
      style={style}
      disabled={disabled}
      title={title}
      value={value}
      type={type}
      component={as!}
    >
      <Container
        as='span'
        inline
        flex
        direction='row'
        alignItems='center'
        className={cx({ [hiddenClass]: loading }, contentClass)}
      >
        {finalChildren}
      </Container>

      {loading && (
        <Loader variant='inherit' className={loaderClass} inline size='small' />
      )}
    </ButtonBase>
  )
}) as CompoundedComponentWithRef<Props, HTMLButtonElement, StaticProps>

Button.defaultProps = {
  active: false,
  as: 'button',
  children: null,
  circular: false,
  disabled: false,
  focused: false,
  fullWidth: false,
  hovered: false,
  iconPosition: 'left',
  loading: false,
  onClick: defaultOnClick,
  size: 'medium',
  type: 'button',
  variant: 'primary-blue'
}

Button.displayName = 'Button'

Button.Group = Group

export default Button as OverridableComponent<Props> & StaticProps
