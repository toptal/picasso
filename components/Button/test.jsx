import React from 'react'
import Button from './index'
import { render, fireEvent, cleanup, waitForElement } from 'react-testing-library'
const Differencify = require('differencify')
const differencify = new Differencify({
  debug: true
})

const renderButton = (children, props = {}) => {
  return render(
    <Button {...props}>{children}</Button>
  )
}

afterEach(cleanup)

test('onClick callback should be fired after clicking the button', () => {
  const onClick = jest.fn()
  const { getByText, container } = renderButton('Click me!', { onClick })

  fireEvent.click(getByText('Click me!'))

  expect(onClick).toHaveBeenCalled()
})

describe('disabled button', () => {
  let onClick
  let api

  beforeEach(() => {
    onClick = jest.fn()

    api = renderButton('Click me!', {
      onClick,
      disabled: true
    })
  })
  test('renders disabled version', () => {
    const { container } = api

    expect(container).toMatchSnapshot()
  })

  test('disables button events', () => {
    const { getByText } = api

    fireEvent.click(getByText('Click me!'))

    expect(onClick).not.toHaveBeenCalled()
  })
})

describe('Visual regression', () => {
  beforeAll(async () => {
    await differencify.launchBrowser({ args: ['--no-sandbox', '--disable-setuid-sandbox'] })
  })

  afterAll(async () => {
    await differencify.cleanup()
  })

  test('validate component', async () => {
    return await differencify
      .init()
      .launch()
      .newPage()
      .goto('http://localhost:9001/?selectedKind=Button&selectedStory=primary&full=0&addons=1&stories=1&panelRight=0')
      .screenshot()
      .toMatchSnapshot()
      .close()
      .end()
  }, 30 * 1000);
});
