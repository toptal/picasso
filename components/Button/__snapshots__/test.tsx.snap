// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`disabled button renders disabled version 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <button
      class="MuiButtonBase-root Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class=""
        style="margin: 0em 0em 0em 0em; display: inline-flex; flex-direction: row; align-items: center;"
      >
        Click me!
      </div>
    </button>
  </div>
</div>
`;
